### DISCUSSION TITLE
Pass by Assignment

### DISCUSSION THREAD

**Discussion by 107602505** [1101-GS4719-C/Python-Community-GS4719-C#165] [Q&A] [answered]
不知道為什麼上面跟下面的outside function 最後出來的結果會不一樣，需要求解XD
![image](https://user-images.githubusercontent.com/90819282/143047818-46823126-5f3e-41c0-8865-9a6e410d94e2.png)

**Comment by EricAgumonJou** [comment, marked as answer]
我也想問這一題😭
我猜是不是因為 int為不可變更資料形式的關係，而list為可變更資料形式。

**Comment by enoch-lightning** [comment]
我想除了可以從物件值(value)可變(mutable)與不可變(immutable)的角度切入，也可以從全域變數與區域變數(函式裡的變數)的角度來討論。

**Comment by ghost** [comment]
https://blog.hitripod.com/python-function-passed-by-assignment/
這篇可以看看文字說明(文中有個明顯筆誤mutable自己忽略一下)，程式部分我電腦顯示的很怪所以我不知道他在寫什麼
或者去看文中推薦的英文討論https://stackoverflow.com/questions/986006/how-do-i-pass-a-variable-by-reference
不去管後面的原因的話，
當Pass mutable object by Assignment ，def內的變數跟def外的同名變數，用print(id())檢查，會發現兩者記憶體位置不同，是兩個不同的物件
當Pass immutable object by Assignment，def內的變數跟def外的同名變數，用print(id())檢查，會發現兩者記憶體位置一樣，是同個物件
跟前面同學回覆的一樣

**Comment by ghost** [comment]
另一篇中文說明 貼上懶人結論
來源https://medium.com/starbugs/python-%E4%B8%80%E6%AC%A1%E6%90%9E%E6%87%82-pass-by-value-pass-by-reference-%E8%88%87-pass-by-sharing-1873a2c6ac46
![1](https://user-images.githubusercontent.com/90567166/143781850-39b21049-769f-4645-923e-d2a32fef2101.jpg)
